quinn-proto/src/connection/streams/recv.rs

```diff
@@ -257,69 +196,173 @@ impl Recv {
+                        ShouldTransmit(false),
+                        max_dirty,
+                    );
+                    self.fused = Some(ReadStatus::Reset(error_code));
+                    ReadStatus::Reset(error_code)
+                }
+                RecvState::Closed => {
+                    self.fused = Some(ReadStatus::Closed);
+                    ReadStatus::Closed
+                }
+            },
+        };
+
+        ReadResult {
+            chunk,
+            status,
+            should_transmit,
+        }
+    }
+}
 
-pub(crate) trait BytesRead {
-    fn bytes_read(&self) -> u64;
+/// Errors that might occur when opening up a reader for a stream
+#[derive(Clone, Copy, Debug, Error, PartialEq, Eq, Hash)]
+pub enum ReadableError {
+    /// Attempted an ordered read following an unordered read
+    ///
+    /// Performing an unordered read allows discontinuities to arise in the receive buffer of a
+    /// stream which cannot be recovered, making further ordered reads impossible.
+    #[error("ordered read after unordered read")]
+    IllegalOrderedRead,
+    /// The stream has not been opened or was already stopped, finished, or reset
+    #[error("unknown stream")]
+    UnknownStream,
 }
 
-impl BytesRead for Chunk {
-    fn bytes_read(&self) -> u64 {
-        self.bytes.len() as u64
+impl From<IllegalOrderedRead> for ReadableError {
+    fn from(_: IllegalOrderedRead) -> Self {
+        ReadableError::IllegalOrderedRead
     }
 }
 
-pub(crate) struct ReadChunks {
-    pub bufs: usize,
-    pub read: usize,
+/// Results of reading from a stream
+#[derive(Debug, PartialEq)]
+#[non_exhaustive]
+pub struct ReadResult {
+    /// The data read from the stream
+    pub chunk: Option<Chunk>,
+    /// The status of the stream after reading
+    pub status: ReadStatus,
+    /// Whether the connection is ready to transmit
+    pub should_transmit: ShouldTransmit,
 }
 
-impl BytesRead for ReadChunks {
-    fn bytes_read(&self) -> u64 {
-        self.read as u64
+impl ReadResult {
+    fn from_status(status: ReadStatus) -> Self {
+        Self {
+            chunk: None,
+            status,
+            should_transmit: ShouldTransmit(false),
+        }
     }
 }
 
-/// Errors triggered when reading from a recv stream
-#[derive(Debug, Error, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
-pub enum ReadError {
-    /// No more data is currently available on this stream.
+/// Status of a receive stream
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd, Hash)]
+pub enum ReadStatus {
+    /// No more data is currently available on this stream
     ///
     /// If more data on this stream is received from the peer, an `Event::StreamReadable` will be
     /// generated for this stream, indicating that retrying the read might succeed.
-    #[error("blocked")]
     Blocked,
-    /// The peer abandoned transmitting data on this stream.
+    /// The stream has been stopped
+    Closed,
```

Comment:
This looks redundant to the `UnknownStream` case in `ReadableError`; a stream can never become stopped mid-read. It's also confusingly named.